connectors/config.py

```diff
@@ -177,3 +180,28 @@ def _merge_dicts(hsh1, hsh2):
             yield (k, hsh1[k])
         else:
             yield (k, hsh2[k])
+
+
+class DataSourceFrameworkConfig:
+    """
+    The configs that will be exposed to DataSource instances.
+    This abstraction prevents DataSource instances from having access to all configuration, while also
+    preventing them from requiring substantial changes to access new configs that may be added.
+    """
+
+    def __init__(self, max_file_size):
+        """
+        Should not be called directly. Use the Builder.
+        """
+        self.max_file_size = max_file_size
+
+    class Builder:
+        def __init__(self):
+            self.max_file_size = DEFAULT_MAX_FILE_SIZE
+
+        def with_max_file_size(self, max_file_size):
+            self.max_file_size = max_file_size
+            return self
+
+        def build(self):
+            return DataSourceFrameworkConfig(self.max_file_size)
```

Comment:
Curious - why use builder here?

For now it's only one property, and probably in future it'll be like 10.

Do you see partial initialisation of this config (e.g. with some properties initialised, some not), or multistep initialisation of config?

Reply:
I want to make it easy for this to be formed up from multiple places eventually. We had a recent slack thread where we discussed how some service configs might one day have a UI and be stored in ES. And we've got some in the YAML, and maybe would someday have some in RCFs. Passing around a half-complete builder is a good pattern to use for scenarios like this where multiple actors are all contributing parts to something being built.

Reply:
Yeah I understand that, though I felt that purpose of Builder is often to have a mutable object until at some point you're ready to create an immutable object and use it somewhere. Since Python has different encapsulation concepts, it's much less necessary, and Builder becomes more of a "convenience" class that makes creation of objects more pleasant.

I'd think that just having `DataSourceFrameworkConfig` class with getters/setters should be fine.

This however should not prevent from merging change as is, it's rather a discussion to see your POV here :)

Reply:
I think the benefit of Builder pattern is that it makes the object instantiation simple and readable when there are a lot of variables. Apparently this is implemented for future-proof.

Do you think we should define the constructor of `DataSourceFrameworkConfig` like this:
```python
def __init__(self, builder):
	self.max_file_size = builder.max_file_size
```
and the `build` method like this:
```python
def build(self):
	return DataSourceFrameworkConfig(self)
```

Reply:
I think the truth is there will be N sources that provide config options, e.g.:

- Configuration file
- Connector-specific config
- Deployment-wide settings

So builder will have 3 args/3 `with_x` methods and number is unlikely to grow IMO, I might be wrong here.

As to the syntax, I think original syntax in Sean's changeset is already good Builder syntax - only improvement (also absolutely unnecessary) could be is to make DataSourceFrameworkConfig have read-only properties with data stored in `__`-prefixed keys so that they could not be accessed, but I don't think it matters a lot.

Reply:
>I think the truth is there will be N sources that provide config options, e.g.:
> 
> - Configuration file
> - Connector-specific config
> - Deployment-wide settings
> 
> So builder will have 3 args/3 with_x methods


I think the builder should not care of where this config is coming from. Take `max_file_size` as example, it should only define one method `with_max_file_size`. If it's from configuration file, you can do `builder.with_max_file_size(loaded_config.max_file_size)`, and if it's from a connector, you can do `builder.with_max_file_size(connector.max_file_size)`

Reply:
Let's just keep these options in mind as this grows. I think we've agreed that the current state is "fine", and that it could be improved on depending on how we use it in the future. So let's just plan to refactor as necessary as its usage grows.